### PR TITLE
Add telemetry support

### PR DESCRIPTION
**Pull request by schugabe**
This PR adds (optional) telemetry support.

Three new components are added:

- Telemetry Receiver
- Stubborn Sender
- Stubborn Receiver

### Telemetry Receiver

This is a rewrite of the crsf telemetry receiver that was already implemented. The receiver buffers the last received telemetry messages and has support to get the next updated messages. The following telemetry messages are supported: GPS, BATTERY_SENSOR, ATTITUDE, DEVICE_INFO, FLIGHT_MODE

Even if the telemetry support is disabled, some parts are required since the bootloader for stm32 targets is called via telemetry messages.

### Stubborn Sender
The sender runs on the receiver on the quad. In HandleSendTelemetryResponse two different types of telemetry packages are sent to the tx: ELRS_TELEMETRY_TYPE_LINK (as before) and ELRS_TELEMETRY_TYPE_DATA. The new message sends 5 bytes at the same time until the tx flips one bit to confirm the reception. Then the next 5 bytes are sent until the bit is flipped again. This simple mechanism ensures that the data isn't duplicated and that the retransmit happens until the reception is confirmed.

If sender/receiver lose sync there is an automatic sync recovery in place.

ELRS_TELEMETRY_TYPE_LINK and ELRS_TELEMETRY_TYPE_DATA is sent in an alternating order. 

- To get the same link update rate the telemetry rate can be set to 32. 
- If you set a telemetry rate of 16, the ELRS_TELEMETRY_TYPE_DATA is sent twice as often. This way the link update happens at the same frequency as in the case of the telemetry rate of 64 but the telemetry data is sent faster

### Stubborn Receiver
The receiver runs on the tx. The confirmation bit is sent on the slot of one of the 1-Bit switches or on the free bit in the hybrid switch data. When the receiver receives one complete frame, it's sent to opentx.

There are native tests for all modules that can be run without any hardware. I extracted some generic crsf definitions to a separte header file so I can use this defines without pulling in hardware dependent includes.

### Test the PR

- Enable telemetry in your FC
- Scan for sensors on TX
- check if you see the voltage RxBt value on the TX to verify that basic telemetry is working
- Download my fork of the repository: https://github.com/schugabe/ExpressLRS 
- Open it in Visual Studio Code
- In bottom left of the IDE you see "master-dev", click on it and select telemetry
- Edit user settings: make sure you have the lines with "-DENABLE_TELEMETRY" and "-DHYBRID_SWITCHES_8" enabled
- Flash TX + RX
- Scan for sensors the list should be longer
- Increase the telemetry rate with the expresslrs lua script. Increase the rate until the sensor lost warnings go away. I use up to 1:16 with 200Hz

### CODE REVIEW

**Comment by AlessandroAU** [comment]
You've done a lot of work! Thank you. I promise I will take a deep look into this soon.

**Comment by schugabe** [comment]
@AlessandroAU I did merge in the current changes to the branch and it seems that everything is still working.

You mentioned on discord that you would like to send the LinkStatistics also with the new method. What would be the advantage? It does not seem that there is any more data in the LinkStatistics used.

**Comment by mdj36** [comment]
Gettinf error on Frsky_RX_R9MM_R9MINI_via_BetafkightPassthrough:
> Executing task: pio run --target upload --environment Frsky_RX_R9MM_R9MINI_via_BetaflightPassthrough <

Processing Frsky_RX_R9MM_R9MINI_via_BetaflightPassthrough (platform: ststm32@8.0.0; board: R9MM; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
UID bytes: 27,62,42,104,52,194
InvalidGitRepositoryError: /home/donatas/ExpressLRS-telemetry/src:
  File "/home/donatas/.platformio/penv/lib/python3.6/site-packages/platformio/builder/main.py", line 168:
    env.SConscript(item, exports="env")
  File "/home/donatas/.platformio/packages/tool-scons/scons-local-4.0.1/SCons/Script/SConscript.py", line 598:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/home/donatas/.platformio/packages/tool-scons/scons-local-4.0.1/SCons/Script/SConscript.py", line 287:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/home/donatas/ExpressLRS-telemetry/src/python/build_flags.py", line 57:
    git_repo = Repo(os.getcwd(), search_parent_directories=True)
  File "/home/donatas/.platformio/penv/lib/python3.6/site-packages/git/repo/base.py", line 181:
    raise InvalidGitRepositoryError(epath)
====================================================================== [FAILED] Took 0.47 seconds ======================================================================

Environment                                     Status    Duration
----------------------------------------------  --------  ------------
Frsky_RX_R9MM_R9MINI_via_BetaflightPassthrough  FAILED    00:00:00.468
================================================================= 1 failed, 0 succeeded in 00:00:00.468 =================================================================
The terminal process "pio 'run', '--target', 'upload', '--environment', 'Frsky_RX_R9MM_R9MINI_via_BetaflightPassthrough'" terminated with exit code: 1.

Terminal will be reused by tasks, press any key to close it.

**Comment by greenbigfrog** [comment]
@mdj36 please join us in the Discord located at http://discord.gg/dS6ReFY for support and we'll try to help you. A Pull Request is definetly the wrong place :)
Also, please make sure to read through the wiki at https://github.com/AlessandroAU/ExpressLRS/wiki

**Comment by mdj36** [comment]
From git clone everything runs go testing gps telemetry in Q7

**Comment by mdj36** [comment]
Sory but no gps telemetry with build: 305f26b

**Comment by IllusionFpv** [comment]
@schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)

**Comment by schugabe** [comment]
> 
> 
> @schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)

If your quad does not output telemetry nothing is sent back to the TX. So you could simply flash everything with telemetry enabled and use the settings of the FC to disable/enable telemetry as you need it. But I have to try this use case before I can really confirm that it works. Did you try my branch? Does it work for you?

**Comment by IllusionFpv** [comment]
@schugabe not tried yet, but I refer to the possibilty of using default basic telemetry on quad and enable telemetry on wing's rx for example

**Comment by ssombra** [comment]
> > @schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)
> 
> If your quad does not output telemetry nothing is sent back to the TX. So you could simply flash everything with telemetry enabled and use the settings of the FC to disable/enable telemetry as you need it. But I have to try this use case before I can really confirm that it works. Did you try my branch? Does it work for you?

How can I try your branch? I would like to try telemetry on my quads.

**Comment by schugabe** [comment]
> 
> 
> > > @schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)
> > 
> > 
> > If your quad does not output telemetry nothing is sent back to the TX. So you could simply flash everything with telemetry enabled and use the settings of the FC to disable/enable telemetry as you need it. But I have to try this use case before I can really confirm that it works. Did you try my branch? Does it work for you?
> 
> How can I try your branch? I would like to try telemetry on my quads.

I updated the description of the PR. If you need more information you could join the discord.

**Comment by schugabe** [comment]
> 
> 
> @schugabe not tried yet, but I refer to the possibilty of using default basic telemetry on quad and enable telemetry on wing's rx for example

The link status telemtry still would be sent back but the battery voltage isn't sent any more. So there is basic telemetry but if you need more data you have the option to enable this feature.

**Comment by ssombra** [comment]
> > > > @schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)
> > > 
> > > 
> > > If your quad does not output telemetry nothing is sent back to the TX. So you could simply flash everything with telemetry enabled and use the settings of the FC to disable/enable telemetry as you need it. But I have to try this use case before I can really confirm that it works. Did you try my branch? Does it work for you?
> > 
> > 
> > How can I try your branch? I would like to try telemetry on my quads.
> 
> I updated the description of the PR. If you need more information you could join the discord.

Just flashed your branch and, although I do receive more sensors, including GPS which was I was looking for, I keep getting "telemetry lost", "telemetry recovered" messages.

**Comment by ssombra** [comment]
> > > > > @schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)
> > > > 
> > > > 
> > > > If your quad does not output telemetry nothing is sent back to the TX. So you could simply flash everything with telemetry enabled and use the settings of the FC to disable/enable telemetry as you need it. But I have to try this use case before I can really confirm that it works. Did you try my branch? Does it work for you?
> > > 
> > > 
> > > How can I try your branch? I would like to try telemetry on my quads.
> > 
> > 
> > I updated the description of the PR. If you need more information you could join the discord.
> 
> Just flashed your branch and, although I do receive more sensors, including GPS which was I was looking for, I keep getting "telemetry lost", "telemetry recovered" messages.

My setup: QX7, RM9 and r9mm

**Comment by schugabe** [comment]
@ssombra I forgot to mention that you have to increase the telemetry rate with the expresslrs lua script. Increase the rate until the warnings go away. I use up to 1:16 with 200Hz to get the fastest update rate. 
If the warnings are not resolved this way please post the Hz/tlm rate settings and also what flight controller + firmware you are using.

**Comment by ssombra** [comment]
> @ssombra I forgot to mention that you have to increase the telemetry rate with the expresslrs lua script. Increase the rate until the warnings go away. I use up to 1:16 with 200Hz to get the fastest update rate.
> If the warnings are not resolved this way please post the Hz/tlm rate settings and also what flight controller + firmware you are using.

That did the trick! thanks. Which is the downside of increasing the telemetry rate? is there a deterioration of the performance in some way? range for instance? I flashed a long range quad I had set up with iNav and of course GPS information is valuable, but if there is a trade off between having that information in the radio and losing range, I can live with only the OSD GPS information.

**Comment by greenbigfrog** [comment]
Basically, at 1:16, it'll send 16 controll packets, pause for one and listen for the telemetry packet, before then sending another 16, waiting for one telemetry, and so on
It's not about range, more about an increase in latency etc again from my understanding since every once in a while you're trading a control packet for a telemtry one.

Your rx only sends out telemetry with roughly 50mW IIRC, so it's your telemetry link limiting your range.
For a dedicated longrange system that requires telemetry (ardupilot etc), you'll want to look into using a seperate telemetry link with higher bidirectional power most likely.

**Comment by ssombra** [comment]
> Basically, at 1:16, it'll send 16 controll packets, pause for one and listen for the telemetry packet, before then sending another 16, waiting for one telemetry, and so on
> It's not about range, more about an increase in latency etc again from my understanding since every once in a while you're trading a control packet for a telemtry one.
> 
> Your rx only sends out telemetry with roughly 50mW IIRC, so it's your telemetry link limiting your range.
> For a dedicated longrange system that requires telemetry (ardupilot etc), you'll want to look into using a seperate telemetry link with higher bidirectional power most likely.

Thanks for the explanation, I think I got it. Actually when I said long range it is more likely medium range cruiser (not expecting going tens of kms away). In both cases anyway, latency is not a big of an issue. Regarding telemetry on the radio, it is just for having coordinates in case of an accident and no OSD recording (it happens :)).

**Comment by Rogero-** [comment]
hi, i know the hard work and complications already present with the Telemetry addition to ELRS, just please keep in mind the Betaflight Configuration LUA script requirements if possible to have it working someday, it helps a great deal in tuning quads and changing settings on the field (maybe as an option that can be enabled or not if it affects the Link performance/latency) since you're the only one currently working on the Telemetry related code, and thanks for your efforts.

**Comment by schugabe** [comment]
I know that some people would like to have the LUA feature but I did not look into it yet.

Since I got some feedback and it seems to work for other people as well I'm removing the draft flag of the PR.

**Comment by Diwellington** [comment]
Waiting for this great feature!!!
Other great value, amazing work 🙌🏽

**Comment by jejakmedia** [comment]
I've tried it. but telemetry works fine at 200hz pack rates

**Comment by jejakmedia** [comment]
> > > > > @schugabe one question about this code, I have some quads where I don't need telemetry but I want it in fixed wings, is it possible to enable telemetry option only in TX? So I can have telemetry only in crafts where I need (flashing with no telemetry in quad rxes for example)
> > > > 
> > > > 
> > > > If your quad does not output telemetry nothing is sent back to the TX. So you could simply flash everything with telemetry enabled and use the settings of the FC to disable/enable telemetry as you need it. But I have to try this use case before I can really confirm that it works. Did you try my branch? Does it work for you?
> > > 
> > > 
> > > How can I try your branch? I would like to try telemetry on my quads.
> > 
> > 
> > I updated the description of the PR. If you need more information you could join the discord.
> 
> Just flashed your branch and, although I do receive more sensors, including GPS which was I was looking for, I keep getting "telemetry lost", "telemetry recovered" messages.

I've tried it. but telemetry works fine at 200hz pack rates

**Comment by bobbillfold** [comment]
@jejakmedia Did you get Betaflight LUA 1.5 scripts working? I have telemetry, but BF LUA stuck on initializing screen

**Comment by greenbigfrog** [comment]
> @jejakmedia Did you get Betaflight LUA 1.5 scripts working? I have telemetry, but BF LUA stuck on initializing screen
Are you actually running this branch/code? (Not sure if it works even if you do.)

**Comment by bobbillfold** [comment]
@jejakmedia I used the RC, but didn't think BF LUA was fully supported yet. Still doing my research, but I think we need full telemetry. Just got that same answer on discord.

**Comment by schugabe** [comment]
This branch does not have full lua script support since I only added messages from the quad back to tx and all the MSP messages (both directions) are missing. There is no technical reason but the PR is large enough and I want wait until this PR gets merged before I continue to work on the remaining features.

File: src/lib/CrsfProtocol/crsf_protocol.h
Comment (greenbigfrog) [inline src/lib/CrsfProtocol/crsf_protocol.h:131]: Any reason why this was left in?

File: src/lib/CrsfProtocol/crsf_protocol.h
Comment (greenbigfrog) [inline src/lib/CrsfProtocol/crsf_protocol.h:252]: Shouldn't this empty line be removed?

File: src/lib/CrsfProtocol/crsf_protocol.h
Comment (greenbigfrog) [inline src/lib/CrsfProtocol/crsf_protocol.h:287]: Can't we remove this old(?) code?

File: src/lib/Telemetry/telemetry.cpp
Comment (greenbigfrog) [inline src/lib/Telemetry/telemetry.cpp:15]:
```suggestion
    telemetry_state = TELEMETRY_IDLE;
```

Any particular reason for this typo? Might just be missing something here in the bigger picture.

Comment (greenbigfrog) [reply src/lib/Telemetry/telemetry.cpp:15]: Should be able to just "find & replace" all places

File: src/lib/Telemetry/telemetry.h
Comment (greenbigfrog) [inline src/lib/Telemetry/telemetry.h:8]:
```suggestion
    RECEIVING_LENGTH,
```

File: src/src/rx_main.cpp
Comment (greenbigfrog) [inline src/src/rx_main.cpp:72]:
```suggestion
StubbornSender TelemetrySender(ELRS_TELEMETRY_MAX_PACKAGES);
```

**Comment by bobbillfold** [comment]
@schugabe Thank you. I'll wait patiently for this and lite pro. 👍

File: src/src/rx_main.cpp
Comment (greenbigfrog) [inline src/src/rx_main.cpp:265]:
```suggestion
                NextTelemetryType = ELRS_TELEMETRY_TYPE_LINK;
```

File: src/src/tx_main.cpp
Comment (greenbigfrog) [inline src/src/tx_main.cpp:165]: I suppose this could be ommited

File: src/user_defines.txt
Comment (greenbigfrog) [inline src/user_defines.txt:67]:
```suggestion
# Comment this to disable advanced telemetry.
-DENABLE_TELEMETRY
```

**Review by greenbigfrog — CHANGES_REQUESTED** [review]
Found some slight typos. We've all been there that we stared at a certain word for too long we've stopped seeing typos.
Otherwise this looks really good. Thanks :smile: 

Perhaps the PR could be switched to target/based on `develop`, to make sure we test it against the latest code before merging. Although I suppose I could rebase it locally.

**Comment by JFly02** [comment]
Does this branch support the fan mod because i enabled the higher power but my fan didn't turn on.

**Comment by greenbigfrog** [comment]
> Does this branch support the fan mod because i enabled the higher power but my fan didn't turn on.

This code doesn't alter any code related to the fan mod AFAIK. Support request of a more basic nature should (IMO) be asked in the discord server https://discord.gg/bARM4WhXFv

**Comment by schugabe** [comment]
> Found some slight typos. We've all been there that we stared at a certain word for too long we've stopped seeing typos.
> Otherwise this looks really good. Thanks smile
> 
> Perhaps the PR could be switched to target/based on `develop`, to make sure we test it against the latest code before merging. Although I suppose I could rebase it locally.

Thank you for the review. Telemetry seems to be a really hard word :rofl: I will test the feature with the recent changes from the develop branch and update the branch with the fixed typos.

**Comment by schugabe** [comment]
Typos are fixed and changes from develop branch are also added to this branch. I did not rebase it again because there are so many conflicts to resolve. Before we merge this feature I can squash the commits (if needed)

**Comment by kara-b** [comment]
Telemetry itself seems to work fine, Not compatible with the Yaapu scripts though as they don't update with any info.

**Comment by jejakmedia** [comment]
i hope gps telemtry can read at telemetry viewer app (android) from R9M module flashed ELRS

**Comment by alexagv** [comment]
Any news on this PR?

**Comment by mdj36** [comment]
> i hope gps telemtry can read at telemetry viewer app (android) from R9M module flashed ELRS

It looks like r9 external module (TX) does not work after express lrs flashed. Need some fix. I tested taranis q7 ->R9 (2019) ->bluetooth Android telemetry app .. original R9 works fine.

**Comment by cruwaller** [comment]
Rebased after PR merges to `develop` branch. Fixes also build issue with 2400 targets.

**Comment by jejakmedia** [comment]
> 
> 
> Rebased after PR merges to `develop` branch. Fixes also build issue with 2400 targets.

where is i can download new file merger?

**Comment by cruwaller** [comment]
> > Rebased after PR merges to `develop` branch. Fixes also build issue with 2400 targets.
> 
> where is i can download new file merger?

Just fetch this PR again :)
You might need to reset your local checkout if you have clone the repo. Or just clone it again :)
I think easiest is to use ELRS Configurator which should handle all automatically :)

**Comment by jejakmedia** [comment]
> 
> 
> > > Rebased after PR merges to `develop` branch. Fixes also build issue with 2400 targets.
> > 
> > 
> > where is i can download new file merger?
> 
> Just fetch this PR again :)
> You might need to reset your local checkout if you have clone the repo. Or just clone it again :)
> I think easiest is to use ELRS Configurator which should handle all automatically :)

do use this clone? https://github.com/schugabe/ExpressLRS

**Comment by jejakmedia** [comment]
> 
> 
> > > Rebased after PR merges to `develop` branch. Fixes also build issue with 2400 targets.
> > 
> > 
> > where is i can download new file merger?
> 
> Just fetch this PR again :)
> You might need to reset your local checkout if you have clone the repo. Or just clone it again :)
> I think easiest is to use ELRS Configurator which should handle all automatically :)

hi, i can't find -DENABLE_TELEMETRY" in new merger?

**Comment by cruwaller** [comment]
Oh my bad, Configurator does not have that new flag :/
You need to use old school method 😎

**Comment by jurgelenas** [comment]
> Oh my bad, Configurator does not have that new flag :/
> You need to use old school method 😎

I could add that user define in the configurator :)

**Comment by jejakmedia** [comment]
> 
> 
> Oh my bad, Configurator does not have that new flag :/
> You need to use old school method sunglasses

Telemetry Issues Latest Clone PR # 204

my setup:  100Hz 100 mw TLM Ratio 1:32 & 1:16 (same result)

https://youtu.be/IlCAHUr6VA0

A new problem after PR (Add telemetry support # 204) merged. In this video I use the last clone PR # 204. The results of the gps telemetry are very slow as is the TBS Crossfire.
In December 2020 I tried clone PR # 204 and the results were satisfactory, GPS telemetry was very fast. Does anyone have a December 2020 clone file? I'll put it back on or anyone can solve this issues?
https://github.com/ExpressLRS/ExpressLRS/pull/204

**Comment by jejakmedia** [comment]
the more telemetry messages, the slower the speed of telemetry such as TBS Crossfire. do can speed telemetry of GPS like as TQly (Like i before rebased)

**Comment by brandonrc** [comment]
Found this in Facebook. Someone posted it. Looks like people are complaining about GPS update speed on telemetry. 

Marios Chrysomallis solve: https://youtu.be/7K3g4HDLw2s

**Comment by brandonrc** [comment]
Nvm saw it on the development branch as well... late to the party

**Comment by expipiplus1** [comment]
I think that the readme could use an update as the mentions of telemetry functionality seem a little out of date now.

**Comment by sfjuocekr** [comment]
I've compiled and flashed the latest develop branch to my newly received HappyModel ES915.

I do get telemetry, but only the first ten values. GPS, BATTERY_SENSOR, ATTITUDE, DEVICE_INFO and FLIGHT_MODE are not received.

**Comment by schugabe** [comment]
> I've compiled and flashed the latest develop branch to my newly received HappyModel ES915.
> 
> I do get telemetry, but only the first ten values. GPS, BATTERY_SENSOR, ATTITUDE, DEVICE_INFO and FLIGHT_MODE are not received.

Did you enable telemetry? You have to remove the # from this line:
https://github.com/ExpressLRS/ExpressLRS/blob/develop/src/user_defines.txt#L66

For more questions please join the discord

**Comment by sfjuocekr** [comment]
Yes, I did remove the comment hash (I wasn't born yesterday).

I think I found the culprit, the TX pin on the FC I tried to use seems to be high all the time. Will try another FC today, but I'm fairly sure that is the issue as I couldn't get passthrough flashing to work on that one either.

**Comment by mdj36** [comment]
At a moment most stable release for me:
https://github.com/schugabe/ExpressLRS/tree/telemetry-before-rebase Yes telemetry update would be nice to have faster (current 2.5 sec update time), but R9 2019 and r9mm working just fine i use 100 mw power for flights around me ( till 1.5 km from me). now issue stable works use it on my favorite 5 drones.
You can have always fresh with many bugs and really unstable: https://github.com/ExpressLRS/ExpressLRS/tree/develop version but flashing twice or more time per day i am little bit tied :)

**Comment by Pairan** [comment]
@schugabe 

Perhaps I'm to new in this topic - but how do I build the Firmware after having it in VSCODE?

**Comment by sergeyprokofev** [comment]
> I've compiled and flashed the latest develop branch to my newly received HappyModel ES915.
> 
> I do get telemetry, but only the first ten values. GPS, BATTERY_SENSOR, ATTITUDE, DEVICE_INFO and FLIGHT_MODE are not received.
Confirm. Inav only has the first ten telemetry values. In BF, all values are. Tested on Matek F405se.

**Comment by Pairan** [comment]
@schugabe 
In my tests I found, that the telemetry values first show up after the second start of the flying thing! The Receiver reports to receive telemetry data yet the desired GPS and stuff is missing at that point. 
After shutting down the wing and restarting is, the telemetry immediately shows up and GPS is broadcasted after having a satelite link

... hope this observation helps who is trying this PR too :) Looking forward to see this stable as it can't be something big and nasty (otherwise it wouldn't work, right?)

**Pull request merged**
